webapp/utils/aws/cloudwatch.py

```diff
@@ -0,0 +1,50 @@
+import logging
+from typing import TYPE_CHECKING
+
+import boto3
+from attrs import define, field
+
+if TYPE_CHECKING:
+    from mypy_boto3_logs.client import CloudWatchLogsClient as CloudWatchLogsClientType
+
+from webapp.config import Config
+
+logger = logging.getLogger(__name__)
+
+
+@define
+class CloudWatchLogsClient:
+    """ECS Client for running and monitoring task runs."""
+
+    log_group_name: str = field(
+        factory=lambda: Config().ALMA_SAP_INVOICES_CLOUDWATCH_LOG_GROUP
+    )
+    log_stream_name_prefix: str = field(
+        factory=lambda: f"sapinvoices/{Config().ALMA_SAP_INVOICES_CLOUDWATCH_LOG_GROUP}/"
+    )
+
+    @property
+    def client(self) -> "CloudWatchLogsClientType":
+        return boto3.client("logs")
+
+    def get_logs(self, task_id: str) -> list[str] | None:
+        logger.info("Retrieving CloudWatch logs for task.")
+        log_events = []
+        params = {
+            "logGroupName": self.log_group_name,
+            "logStreamName": f"{self.log_stream_name_prefix}{task_id}",
+            "startFromHead": True,
+        }
+
+        while True:
+            response = self.client.get_log_events(**params)  # type: ignore[arg-type]
```

Comment:
What happens here if the log group/stream does not exist?  The [documentation here](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/logs/client/get_log_events.html#) suggests that `CloudWatchLogs.Client.exceptions.ResourceNotFoundException` can get thrown if that's the case.

Would it be worth considering adding a check _before_ this `while` loop to ensure the log stream exists before attemping to retrieve lines?  Could even be a method like `log_stream_exists()`?  

Or... perhaps this work would come later when integrating into the flow of things?  I could a see a universe where we kickoff the ECS task, but the log stream hasn't _yet_ been created, but we are immediately querying for it.

Maybe don't need to solve now, just noting.

Reply:
I decided to:

[1. Write a method to check if a log stream exists by doing performing an "exact match" of a given task ID against the task IDs retrieved from the log stream names retrieved by `CloudWatchLogsClient.get_log_streams`](https://github.com/MITLibraries/alma-sapinvoices-ui/pull/12/files#diff-5e939bc3e5581f9188452ccbff87a72ed1bd0b78a11a34df843f05f262961248R67-R71)

[2. Use the method from 1 in `CloudWatchLogsClient.get_log_events` and raise a custom exception `ECSTaskLogStreamDoesNotExistError` when a log stream for the task ID is not found](https://github.com/MITLibraries/alma-sapinvoices-ui/pull/12/files#diff-5e939bc3e5581f9188452ccbff87a72ed1bd0b78a11a34df843f05f262961248R54-R65). 

**Note:** I opted to create a custom exception as it allows our exceptions to have consistent naming and write cleaner messages. In the case of the `ResourceNotFoundException`, I would've had to write the `try-except` clause with the following syntax:

```
try;
   <try this>
except self.client.exceptions.ResourceNotFoundException as exception:
   <do this>
```

Because the exception needs to be imported _from_ the `boto3` client, it feels a little confusing within the `CloudWatchLogsClient`. 

Let me know what you think!